File: demo/ctpping/ctpping.cpp
#include <iostream>
#include <chrono>
#include <thread>
#ifdef _WIN32
#include "win/getopt.h"
#else
#include <unistd.h>
#endif
#include "ThostFtdcMdApi.h"
#include "ThostFtdcTraderApi.h"

auto reqtime = std::chrono::steady_clock::now();

class CMarketSpi :public CThostFtdcMdSpi
{
public:
	CMarketSpi(CThostFtdcMdApi* pApi):m_pApi(pApi)
	{
		pApi->RegisterSpi(this);
	}

	void OnFrontConnected()
	{
		std::cout << "connected." << std::endl;

		reqtime = std::chrono::steady_clock::now();
		CThostFtdcReqUserLoginField Req = {0};
		m_pApi->ReqUserLogin(&Req, 0);
	}

	void OnFrontDisconnected(int nReason)
	{
		std::cout << "disconnected." << nReason << std::endl;
		exit(0);
	}

	void OnRspUserLogin(CThostFtdcRspUserLoginField* pRspUserLogin, CThostFtdcRspInfoField* pRspInfo, int nRequestID, bool bIsLast)
	{
		auto rsptime = std::chrono::steady_clock::now();
		auto duration = std::chrono::duration_cast<std::chrono::milliseconds>(rsptime - reqtime);

		std::cout << "response time: " << duration.count() << " milliseconds." << std::endl;
		exit(0);
	}

	CThostFtdcMdApi* m_pApi;
};

class CTradeSpi :public CThostFtdcTraderSpi
{
public:
	CTradeSpi(CThostFtdcTraderApi* pApi) :m_pApi(pApi)
	{
		pApi->RegisterSpi(this);
	}

	void OnFrontConnected()
	{
		std::cout << "connected." << std::endl;

		reqtime = std::chrono::steady_clock::now();
		CThostFtdcReqUserLoginField Req = { 0 };
		m_pApi->ReqUserLogin(&Req, 0);
	}

	void OnFrontDisconnected(int nReason)
	{
		std::cout << "disconnected." << std::endl;
		exit(0);
	}

	void OnRspUserLogin(CThostFtdcRspUserLoginField* pRspUserLogin, CThostFtdcRspInfoField* pRspInfo, int nRequestID, bool bIsLast)
	{
		auto rsptime = std::chrono::steady_clock::now();
		auto duration = std::chrono::duration_cast<std::chrono::milliseconds>(rsptime - reqtime);

		std::cout << "response time: " << duration.count() << " milliseconds" << std::endl;
		exit(0);
	}

	CThostFtdcTraderApi* m_pApi;
};

void print_usage()
{
	std::cout << "usage:ctpping [-s milliseconds] [-t] [-m] address" << std::endl;
	std::cout << "example:ctpping tcp://180.168.146.187:10130" << std::endl;
	std::cout << "example:ctpping -m tcp://180.168.146.187:10131" << std::endl;
	std::cout << "example:ctpping -s 1000 -t tcp://180.168.146.187:10130" << std::endl;
}
int main(int argc,char *argv[])
{
	bool use_trade = false;
	int ch;
	int milliseconds = 3000; // 3 seconds
	while ((ch = getopt(argc, argv, "s:tm")) != -1)
	{
		switch (ch) {
		case 's':
			milliseconds = atol(optarg);
			break;
		case 't':
			use_trade = true;
			break;
		case 'm':
			break;
		case '?':
			print_usage();
			return -1;
		default:
			print_usage();
			return -1;
		}
	}

	if (optind >= argc) {
		print_usage();
		return -1;
	}

	if (use_trade) {
		//std::cout << "version:" << CThostFtdcTraderApi::GetApiVersion() << std::endl;

		CThostFtdcTraderApi* pApi = CThostFtdcTraderApi::CreateFtdcTraderApi();
		CTradeSpi Spi(pApi);
		pApi->RegisterFront(argv[optind]);
		pApi->Init();
		std::this_thread::sleep_for(std::chrono::milliseconds(milliseconds));
	}
	else {
		//std::cout << "version:" << CThostFtdcMdApi::GetApiVersion() << std::endl;

		CThostFtdcMdApi* pApi = CThostFtdcMdApi::CreateFtdcMdApi();
		CMarketSpi Spi(pApi);
		pApi->RegisterFront(argv[optind]);
		pApi->Init();
		std::this_thread::sleep_for(std::chrono::milliseconds(milliseconds));
	}

	std::cout << "time out." << std::endl;

	return 0;
}

File: ctpapi-python/demo/ctpping.py
# author: Jedore https://github.com/Jedore
# date: 2023/07/15
# 描述: ctp通讯链路检测工具；参考c++ ctpping https://github.com/openctp/openctp/tree/master/demo/ctpping
#   使用前安装 openctp-ctp: pip install openctp-ctp==6.6.9
#   适用 Win/Linux/Mac, python 3.7~3.11
# 用法: python ctpping.py address
# 示例: python ctpping.py tcp://180.168.146.187:10131
# 正常输出:
#   version: v6.6.9_MacOS_20220926 15:00:00
#   connected.
#   response time: 27 milliseconds
# 异常输出:
#   version: v6.6.9_MacOS_20220926 15:00:00
#   time out.

import os
import sys
import time
from queue import Queue, Empty

from openctp_ctp import mdapi

Q_EXIT = Queue(maxsize=1)


class CMarketSpi(mdapi.CThostFtdcMdSpi):
    def __init__(self, _api: mdapi.CThostFtdcMdApi):
        super().__init__()

        self._api = _api
        self._start_time = None
        self._end_time = None

    def OnFrontConnected(self):
        print("connected.")

        self._start_time = int(time.time() * 1000)
        self._api.ReqUserLogin(mdapi.CThostFtdcReqUserLoginField(), 0)

    def OnFrontDisconnected(self, nReason: int):
        print("disconnected.")
        Q_EXIT.put_nowait(None)

    def OnRspUserLogin(
        self,
        pRspUserLogin: mdapi.CThostFtdcRspUserLoginField,
        pRspInfo: mdapi.CThostFtdcRspInfoField,
        nRequestID: int,
        bIsLast: bool,
    ):
        self._end_time = int(time.time() * 1000)
        print("response time:", self._end_time - self._start_time, "milliseconds")
        Q_EXIT.put_nowait(None)


if __name__ == "__main__":
    if len(sys.argv) != 2:
        print(sys.argv)
        print("usage: python ctpping.py address")
        print("example: python ctpping.py tcp://180.168.146.187:10100")
        exit(0)

    print("version:", mdapi.CThostFtdcMdApi.GetApiVersion())
    api = mdapi.CThostFtdcMdApi.CreateFtdcMdApi()
    spi = CMarketSpi(api)
    api.RegisterFront(sys.argv[1])
    api.RegisterSpi(spi)
    api.Init()

    try:
        Q_EXIT.get(timeout=5)
    except Empty:
        print("time out.")
    else:
        os.remove("DialogRsp.con")
        os.remove("QueryRsp.con")
        os.remove("TradingDay.con")
    finally:
        api.Release()
